File: modules/aeris-native-calls/android/src/main/java/ru/aeris/nativecalls/AerisNativeCallsModule.kt
package ru.aeris.nativecalls

import android.content.Context
import com.google.firebase.messaging.FirebaseMessaging
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class AerisNativeCallsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("AerisNativeCalls")

    Events(
      "onPushToken",
      "onIncomingCall",
      "onAnswerCall",
      "onEndCall",
      "onSetActive",
      "onSetInactive",
      "onNativeCallError"
    )

    OnCreate {
      AerisNativeCallsManager.initialize(context)
      val sink = { event: String, body: Map<String, Any?> ->
        try {
          sendEvent(event, body)
        } catch (_: Throwable) {
          // The native call remains valid even if React Native is not mounted yet.
        }
      }
      AerisNativeCallsManager.setEventSink(sink)
      AerisNativeCallsModuleEvents.sink = sink
    }

    OnDestroy {
      AerisNativeCallsManager.setEventSink(null)
      AerisNativeCallsModuleEvents.sink = null
    }

    Function("getPendingCallAction") {
      AerisNativeCallsManager.consumePendingAction()
    }

    AsyncFunction("getPushTokenAsync") { promise: Promise ->
      val stored = context.getSharedPreferences(
        AerisFirebaseMessagingService.PREFERENCES,
        Context.MODE_PRIVATE
      ).getString(AerisFirebaseMessagingService.KEY_FCM_TOKEN, null)

      try {
        FirebaseMessaging.getInstance().token
          .addOnSuccessListener { token ->
            context.getSharedPreferences(
              AerisFirebaseMessagingService.PREFERENCES,
              Context.MODE_PRIVATE
            ).edit().putString(AerisFirebaseMessagingService.KEY_FCM_TOKEN, token).apply()
            promise.resolve(token)
          }
          .addOnFailureListener { error ->
            if (stored != null) {
              promise.resolve(stored)
            } else {
              promise.reject("PUSH_TOKEN_UNAVAILABLE", error.message, error)
            }
          }
      } catch (error: Throwable) {
        if (stored != null) {
          promise.resolve(stored)
        } else {
          promise.reject("FIREBASE_NOT_CONFIGURED", error.message, error)
        }
      }
    }

    AsyncFunction("reportIncomingCallAsync") {
      callId: String, callerName: String, hasVideo: Boolean ->
      AerisNativeCallsManager.reportIncomingCall(callId, callerName, hasVideo)
    }

    Function("reportAnswerResult") { callId: String, success: Boolean ->
      AerisNativeCallsManager.reportAnswerResult(callId, success)
    }

    Function("reportCallEnded") { callId: String, failed: Boolean ->
      AerisNativeCallsManager.reportCallEnded(callId, failed)
    }

    Function("requestEndCall") { callId: String ->
      AerisNativeCallsManager.requestEndCall(callId)
    }
  }

  private val context: Context
    get() = checkNotNull(appContext.reactContext) { "React context is unavailable" }
}

File: modules/aeris-native-calls/ios/AerisNativeCallsModule.swift
import ExpoModulesCore
import Foundation

public final class AerisNativeCallsModule: Module, @unchecked Sendable {
  private var observers: [NSObjectProtocol] = []

  public func definition() -> ModuleDefinition {
    Name("AerisNativeCalls")

    Events(
      "onVoipToken",
      "onAnswerCall",
      "onEndCall",
      "onAudioSessionActivated",
      "onAudioSessionDeactivated"
    )

    OnCreate {
      AerisNativeCallsManager.shared.start()
      self.installObservers()
    }

    OnDestroy {
      self.removeObservers()
    }

    Function("getVoipToken") {
      AerisNativeCallsManager.shared.currentVoipToken()
    }

    Function("getPendingCallAction") {
      AerisNativeCallsManager.shared.consumePendingAction()
    }

    AsyncFunction("reportIncomingCallAsync") {
      (callId: String, callerName: String, hasVideo: Bool, promise: Promise) in
      AerisNativeCallsManager.shared.reportIncomingCall(
        callId: callId,
        callerName: callerName,
        hasVideo: hasVideo
      ) { error in
        if let error {
          promise.reject("ERR_NATIVE_CALL", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      }
    }.runOnQueue(.main)

    AsyncFunction("reportAnswerResult") { (callId: String, success: Bool) in
      AerisNativeCallsManager.shared.fulfillAnswer(callId: callId, success: success)
    }.runOnQueue(.main)

    AsyncFunction("reportCallEnded") { (callId: String, failed: Bool) in
      AerisNativeCallsManager.shared.reportCallEnded(callId: callId, failed: failed)
    }.runOnQueue(.main)

    AsyncFunction("requestEndCall") { (callId: String) in
      AerisNativeCallsManager.shared.requestEndCall(callId: callId)
    }.runOnQueue(.main)
  }

  private func installObservers() {
    guard observers.isEmpty else { return }
    let center = NotificationCenter.default

    observers.append(
      center.addObserver(forName: .aerisVoipTokenChanged, object: nil, queue: .main) {
        [weak self] notification in
        let token = notification.userInfo?["token"]
        self?.sendEvent("onVoipToken", ["token": token ?? NSNull()])
      }
    )
    observers.append(
      center.addObserver(forName: .aerisNativeCallAnswered, object: nil, queue: .main) {
        [weak self] notification in
        guard let callId = notification.userInfo?["callId"] as? String else { return }
        self?.sendEvent("onAnswerCall", ["callId": callId])
      }
    )
    observers.append(
      center.addObserver(forName: .aerisNativeCallEnded, object: nil, queue: .main) {
        [weak self] notification in
        guard let callId = notification.userInfo?["callId"] as? String else { return }
        self?.sendEvent("onEndCall", ["callId": callId])
      }
    )
    observers.append(
      center.addObserver(forName: .aerisCallAudioActivated, object: nil, queue: .main) {
        [weak self] _ in
        self?.sendEvent("onAudioSessionActivated")
      }
    )
    observers.append(
      center.addObserver(forName: .aerisCallAudioDeactivated, object: nil, queue: .main) {
        [weak self] _ in
        self?.sendEvent("onAudioSessionDeactivated")
      }
    )
  }

  private func removeObservers() {
    let center = NotificationCenter.default
    observers.forEach { center.removeObserver($0) }
    observers.removeAll()
  }
}
